scripts/usb_scan.py

```diff
@@ -0,0 +1,618 @@
+
+        if one of device's interfaces is denied by rule, the device is denied.
+        Normally, there should be a catch all rule at last line of policy file:
+            ALLOW: # Otherwise allow everything else
+        If there isn't, and no matching rule is found, the interface will also
+        be denied.
+
+        Only if all interfaces are allowed, the device is allowed to passed
+        through.
+
+        Example of a device before checking
+        UsbDevice: (u'1-2', {'idVendor': '17ef', 'ID_VENDOR_FROM_DATABASE':
+        u'Lenovo', 'bConfigurationValue': '1', 'bDeviceClass': '00',
+        'ID_MODEL_FROM_DATABASE': '', 'version': ' 2.00', 'idProduct': '383c',
+          'bNumInterfaces': ' 1', 'serial': 'H8RU2TGE', 'bcdDevice': '0100'})
+            |__ UsbInterface: (u'1-2:1.0', {'bInterfaceNumber': '00',
+            'bInterfaceClass': '08', 'bInterfaceProtocol': '50',
+            'bInterfaceSubClass': '06'})
+            |__ Allow: False
+
+        UsbDevice: (u'4-1.2', {'idVendor': '2717', 'ID_VENDOR_FROM_DATABASE':
+         '' , 'bConfigurationValue': '1', 'bDeviceClass': '00',
+        'ID_MODEL_FROM_DATABASE': '', 'version': ' 2.00', 'idProduct': 'ff48',
+          'bNumInterfaces': ' 2', 'serial': '7b06864', 'bcdDevice': '0404'})
+            |__ UsbInterface: (u'4-1.2:1.0', {'bInterfaceNumber': '00',
+            'bInterfaceClass': 'ff', 'bInterfaceProtocol': '00',
+            'bInterfaceSubClass': 'ff'})
+            |__ UsbInterface: (u'4-1.2:1.1', {'bInterfaceNumber': '01',
+            'bInterfaceClass': 'ff', 'bInterfaceProtocol': '01',
+            'bInterfaceSubClass': '42'})
+            |__ Allow: False
+
+        :param device:(UsbDevice) the device to check
+        :return:(bool) if allow pass through
+        """
+        log_wrapper("policy check: " + device.get_node())
+        for i in device.get_all_interfaces():
+            allow_interface = False
+            for r in self.rule_list:
+                if self.match_device_interface(r, device, i):
+                    if r[self._ALLOW]:
+                        log_wrapper("allow " + i.get_node())
+                        allow_interface = True
+                        break
+                    else:
+                        log_wrapper("deny " + i.get_node())
+                        return False
+            if not allow_interface:
+                log_wrapper("deny " + i.get_node() + ", no matching rule")
+                return False
+
+        log_wrapper("allow " + device.get_node())
+        return True
+
+
+def parse_args():
+    parser = argparse.ArgumentParser(
+        description="scanner to get USB devices info")
+    parser.add_argument("-d", "--diagnostic", dest="diagnostic",
+                        action="store_true",
+                        help="enable diagnostic mode")
+    return parser.parse_args()
+
+
+def to_pusb(device):
+    """ convert UsbDevice to pusb dict
+
+    Example pusb dict:
```

Comment:
Can we have this in an easier to read format? If this is a dictionary, one entry per line would be much easier to read.

Reply:
done.